packages/langium/src/test/langium-test.ts

```diff
@@ -188,3 +189,123 @@ function replaceIndices(base: ExpectedBase): { output: string, indices: number[]
 
     return { output: input, indices, ranges: ranges.sort((a, b) => a[0] - b[0]) };
 }
+
+export interface ValidationResult<T extends AstNode = AstNode> {
+    diagnostics: Diagnostic[];
+    document: LangiumDocument<T>;
+}
+
+export function validationHelper<T extends AstNode = AstNode>(services: LangiumServices): (input: string) => Promise<ValidationResult<T>> {
+    const parse = parseHelper<T>(services);
+    return async (input) => {
+        const document = await parse(input);
+        return { document, diagnostics: await services.validation.DocumentValidator.validateDocument(document) };
+    };
+}
+
+export type ExpectDiagnosticOptionsWithoutContent<T extends AstNode = AstNode> = ExpectDiagnosticCode & (ExpectDiagnosticAstOptions<T> | ExpectDiagnosticRangeOptions | ExpectDiagnosticOffsetOptions);
+export type ExpectDiagnosticOptions<T extends AstNode = AstNode> = ExpectDiagnosticContent & ExpectDiagnosticOptionsWithoutContent<T>;
+
+export interface ExpectDiagnosticContent {
+    message?: string | RegExp
+    severity?: DiagnosticSeverity;
+}
+
+export interface ExpectDiagnosticCode {
+    code?: string;
+}
+
+export interface ExpectDiagnosticAstOptions<T extends AstNode> {
+    node: T;
+    property?: { name: Properties<T>, index?: number };
+}
+
+export interface ExpectDiagnosticRangeOptions {
+    range: Range;
+}
+
+export interface ExpectDiagnosticOffsetOptions {
+    offset: number
+    length: number
+}
+
+export type Predicate<T> = (arg: T) => boolean;
+
+function isLeftOrEqualFrom(lhs: Position, rhs: Position) {
+    return lhs.line < rhs.line || lhs.line === rhs.line && lhs.character <= rhs.character;
+}
+
+function containsRange(outer: Range, inner: Range): boolean {
+    const start = isLeftOrEqualFrom(outer.start, inner.start);
+    const end = isLeftOrEqualFrom(inner.end, outer.end);
+    return start && end;
+}
```

Comment:
Suggestion: I would replace the `containsRange` function with a simple `isRangeEqual` function. The issue is that we can't safely rely on range, if the node in question (or range, or offset) contains another node with a validation error. For example in the domain model example, we could have something like this:

```
entity A { // We want to check for a validation error on this element
  entity B { // But the check also completes successfully, if the error appears on this element
  }
}
```

This is the reason why we need the `property` property of the interface in the first place; to correctly identify the **exact** range of the diagnostic.

Reply:
I could not find any deep comparison utility, so I implemented it.